### PR TITLE
feat(uniswap_v4): implement `ProtocolSim` except `delta_transition`

### PR DESCRIPTION
**Pull request by zizou0x**
`delta_transition` will be implemented in a future PR because it depends on the Substreams.

### CODE REVIEW

**Review by tvinagre — APPROVED** [review]

File: src/evm/protocol/uniswap_v4/state.rs [hunk 1]
```diff
@@ -217,50 +227,92 @@ impl UniswapV4State {
     }
 }
 
-#[allow(unused_variables)] //TODO: remove when implemented
 impl ProtocolSim for UniswapV4State {
+    // Not possible to implement correctly with the current interface because we need to know the
+    // swap direction.
```
Comment (louise-poole) [inline src/evm/protocol/uniswap_v4/state.rs:232]: Are we going to not implement it at all? Where is it used? If needed we can return some sort of best guess - like a minimum fee?

**Review by louise-poole — APPROVED** [review]

File: src/evm/protocol/uniswap_v4/state.rs (hunk 1 above)
Comment (zizou0x) [reply src/evm/protocol/uniswap_v4/state.rs:232]: It's not used AFAIK, this one isn't implemented for VMs also. We need to find a way to make this interface better, but in the scope of this epic it's not planned to be implemented no. 

The minimum fee was also suggested by Alan, but I find it unexpected from a client POV so I decided to leave it todo for now. 
We can easily come back to this later when they launch, for example if they don't enable any protocol fee (similar to uniswapv3) then we can return lp fee there.

Comment (dianacarvalho1) [reply src/evm/protocol/uniswap_v4/state.rs:232]: yes, fees are not supported in this simulation release.. we need to come back to it at some point

File: src/evm/protocol/uniswap_v4/state.rs [hunk 2]
```diff
@@ -217,50 +227,92 @@ impl UniswapV4State {
...
+        token_in: &Token,
+        token_out: &Token,
     ) -> Result<GetAmountOutResult, SimulationError> {
-        todo!()
+        let zero_for_one = token_in < token_out;
+        let amount_specified = I256::checked_from_sign_and_abs(
+            Sign::Positive,
+            U256::from_be_slice(&amount_in.to_bytes_be()),
+        )
+        .unwrap();
```
Comment (dianacarvalho1) [inline src/evm/protocol/uniswap_v4/state.rs:261]: can you use expect instead of unwrap here? 👀

**Review by dianacarvalho1 — APPROVED** [review]
Thank you @zizou0x ! tiny tiny comment only!

File: src/evm/protocol/uniswap_v4/state.rs (hunk 2 above)
Comment (louise-poole) [reply src/evm/protocol/uniswap_v4/state.rs]: So proud of you, @dianacarvalho1 ❤️ 🤣

Comment (dianacarvalho1) [reply src/evm/protocol/uniswap_v4/state.rs]: I've levelled up in my Rust track!! ➕

**Pull request merged**